incubator/hnc/internal/validators/namespace.go

```diff
@@ -64,19 +64,15 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
 // handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
 // tested (ie without constructing a full admission.Request).
 func (v *Namespace) handle(req *nsRequest) admission.Response {
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
 	parent := req.ns.Annotations[api.SubnamespaceOf]
-	if parent != "" && req.op == v1beta1.Delete {
+	if parent != "" && v.Forest.Get(parent).HasAnchor(req.ns.Name) && req.op == v1beta1.Delete {
 		msg := fmt.Sprintf("The namespace %s is a subnamespace. Please delete the anchor from the parent namespace %s to delete the subnamespace.", req.ns.Name, parent)
 		return deny(metav1.StatusReasonForbidden, msg)
 	}
 
-	return v.checkForest(req)
-}
-
-func (v *Namespace) checkForest(req *nsRequest) admission.Response {
-	v.Forest.Lock()
-	defer v.Forest.Unlock()
-
```

Comment:
This function is already much too long IMO, I don't think we should be making it longer. What's the motivation behind merging them?

If anything, maybe hold off on this change, submit another PR to restructure this webhook so each function is much smaller (maybe max 40 lines) and then reapply this change?

Reply:
Ah the only reason I did so is because we need to check (lock) the forest for that part too now before `checkForest()` func.
sgtm, I will work on the restructure first then come back to this PR.

Reply:
Done in #870